hack/apidiff.sh

```diff
@@ -79,6 +79,33 @@ while getopts "r:t:b:" o; do
 done
 shift $((OPTIND - 1))
 
+# default from prow env if unset from args
+# https://docs.prow.k8s.io/docs/jobs/#job-environment-variables
+# TODO: handle batch PR testing
+
+if [[ -z "${target:-}" && -n "${PULL_PULL_SHA:-}" ]]; then
+    target="${PULL_PULL_SHA}"
+fi
+# target must be a something that git can resolve to a commit.
+# "git rev-parse --verify" checks that and prints a detailed
+# error.
+if [[ -n "${target}" ]]; then
+    target="$(git rev-parse --verify "${target}")"
+fi
+
+if [[ -z "${base}" && -n "${PULL_BASE_SHA:-}" && -n "${PULL_PULL_SHA:-}" ]]; then
+    if ! base="$(git merge-base "${PULL_BASE_SHA}" "${PULL_PULL_SHA}")"; then
+        echo >&2 "Failed to detect base revision correctly with prow environment variables."
+        exit 1
+    fi
+elif [[ -z "${base}" ]]; then
+    if ! base="$(git merge-base origin/master "${target:-HEAD}")"; then
+        echo >&2 "Could not determine default base revision. -r must be used explicitly."
+        exit 1
+    fi
```

Comment:
spotted by @liggitt  in https://github.com/kubernetes/kubernetes/pull/129519#issuecomment-2585032018 , @BenTheElder it seems this detection does not work for cherry picks

Reply:
I suppose it fails because `origin/master` is not the target branch.

`${PULL_BASE_SHA:-origin/master}` should work as replacement of `origin/master`.

Reply:
Scratch that. The actual problem is that `pull-kubernetes-apidiff-client-go` is defined also for release branches which don't have this change yet.

We can either:
- update the release branches,
- maintain per-branch variants of the `pull-kubernetes-apidiff-client-go` job, or
- don't run apidiff for release branches.

I prefer updating the branches. That also gives us an opportunity to lock the the test build against the controller-runtime to a version of that runtime that is compatible with the release branch and do other per-release changes, if the need arises.

@aojea: agreed?

Reply:
seems a non-risk backport, the job is already failing and the code is in the hack directory +1